# Exploratory Data Analysis (EDA)

## Question 1:
How have global layoff trends evolved from 2020–2026?

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [4]:
df = pd.read_csv(
    "../data/processed/layoffs_cleaned.csv",
    parse_dates=["date", "date_added"]
)

df.shape

(4428, 15)

In [5]:
layoff_events_by_year = df.groupby("year").size()

layoff_events_by_year

year
2020     635
2021      44
2022    1230
2023    1387
2024     633
2025     331
2026     168
dtype: int64

In [6]:
df[df["year"] == 2021].head()

,company,location,total_laid_off,date,percentage_laid_off,industry,source,stage,funds_raised,country,date_added,year,month,month_name,quarter
3749,Delivery Hero,"Berlin, Non-U.S.",300.0,2021-12-22,NaN,Food,https://www.businessinsider.de/gruenderszene/f...,Post-IPO,8300.0,Germany,2021-12-30,2021,12,December,4
3750,iFit,Logan,NaN,2021-12-08,NaN,Fitness,https://www.hjnews.com/news/business/ifit-lays...,Private Equity,200.0,United States,2022-11-22,2021,12,December,4
3751,Better.com,New York City,900.0,2021-12-01,0.09,Real Estate,https://techcrunch.com/2021/12/01/better-com-l...,Unknown,905.0,United States,2021-12-01,2021,12,December,4
3752,BitTitan,Seattle,70.0,2021-11-18,0.27,Data,https://www.seattletimes.com/business/seventy-...,Acquired,46.0,United States,2021-11-20,2021,11,November,4
3753,Zillow,Seattle,2000.0,2021-11-02,0.25,Real Estate,https://www.geekwire.com/2021/zillow-shutter-h...,Post-IPO,97.0,United States,2021-11-02,2021,11,November,4


In [7]:
df[df["year"] == 2021].shape

(44, 15)

In [8]:
layoffs_by_year = df.groupby("year")["total_laid_off"].sum()

layoffs_by_year

year
2020     80998.0
2021     15823.0
2022    165269.0
2023    264320.0
2024    152922.0
2025    124636.0
2026    116137.0
Name: total_laid_off, dtype: float64

In [9]:
avg_layoff_size = df.groupby("year")["total_laid_off"].mean()

avg_layoff_size

year
2020     170.522105
2021     510.419355
2022     201.302071
2023     312.434988
2024     394.128866
2025     549.057269
2026    1036.937500
Name: total_laid_off, dtype: float64

In [10]:
quarterly_events = df.groupby(["year", "quarter"]).size()

quarterly_events

year  quarter
2020  1          121
      2          429
      3           61
      4           24
2021  1           22
      2            6
      3            9
      4            7
2022  1           35
      2          304
      3          413
      4          478
2023  1          586
      2          313
      3          276
      4          212
2024  1          240
      2          156
      3          126
      4          111
2025  1          105
      2           72
      3           86
      4           68
2026  1           89
      2           79
dtype: int64

In [11]:
year_summary = pd.DataFrame({
    "Layoff_Events": df.groupby("year").size(),
    "Employees_Laid_Off": df.groupby("year")["total_laid_off"].sum(),
    "Average_Layoff_Size": df.groupby("year")["total_laid_off"].mean()
})

year_summary

,Layoff_Events,Employees_Laid_Off,Average_Layoff_Size
year,,,
2020,635,80998.0,170.522105
2021,44,15823.0,510.419355
2022,1230,165269.0,201.302071
2023,1387,264320.0,312.434988
2024,633,152922.0,394.128866
2025,331,124636.0,549.057269
2026,168,116137.0,1036.937500


# Key Insights

## 1. Layoff activity peaked in 2023.

2023 recorded the highest number of layoff events (1,387) and the highest number of employees laid off (264,320).
This indicates that 2023 was the most severe year for workforce reductions in the dataset.

## 2. Two major layoff waves are visible.

The first wave occurred in 2020, driven by the COVID-19 pandemic and economic uncertainty.
The second and larger wave occurred during 2022–2023, likely reflecting the post-pandemic market correction, funding slowdown, and restructuring across industries.

## 3. Layoff activity declined after 2023.

The number of layoff events fell from 1,387 in 2023 to 633 in 2024.
This suggests that workforce reduction activity began stabilizing after the peak period.

## 4. Average layoff size increased over time.

Average layoff size increased from approximately 171 employees per event in 2020 to over 312 employees per event in 2023.
Although later years show even higher averages, these years may be incomplete and should be interpreted cautiously.

## 5. The largest concentration of layoff events occurred in Q1 2023.

Q1 2023 alone recorded 586 layoff events, the highest quarterly value in the dataset.
This suggests that many organizations initiated workforce reductions at the beginning of 2023.

## Question 2
Which industries experienced the highest layoff activity and workforce reductions?

In [12]:
industry_events = df.groupby("industry").size().sort_values(ascending=False)

industry_events.head(15)

industry
Finance           527
Retail            353
Healthcare        327
Other             302
Consumer          301
Transportation    270
Food              250
Marketing         214
Crypto            165
Real Estate       165
Education         162
Security          157
Media             157
Data              138
HR                113
dtype: int64

In [13]:
industry_layoffs = (
    df.groupby("industry")["total_laid_off"]
      .sum()
      .sort_values(ascending=False)
)

industry_layoffs.head(15)

industry
Other             119746.0
Retail            106706.0
Hardware          106157.0
Consumer           96932.0
Finance            67565.0
Transportation     66002.0
Food               52091.0
Healthcare         39244.0
Infrastructure     24835.0
Travel             23740.0
Sales              21905.0
Education          21002.0
Real Estate        20646.0
HR                 16157.0
Marketing          15835.0
Name: total_laid_off, dtype: float64

In [14]:
industry_summary = pd.DataFrame({
    "Layoff_Events": df.groupby("industry").size(),
    "Employees_Laid_Off": df.groupby("industry")["total_laid_off"].sum()
})

industry_summary.sort_values(
    by="Employees_Laid_Off",
    ascending=False
).head(15)

,Layoff_Events,Employees_Laid_Off
industry,,
Other,302,119746.0
Retail,353,106706.0
Hardware,79,106157.0
Consumer,301,96932.0
Finance,527,67565.0
Transportation,270,66002.0
Food,250,52091.0
Healthcare,327,39244.0
Infrastructure,78,24835.0


In [15]:
industry_avg = (
    df.groupby("industry")["total_laid_off"]
      .mean()
      .sort_values(ascending=False)
)

industry_avg.head(15)

industry
Hardware          1930.127273
Other              633.576720
Infrastructure     605.731707
Consumer           489.555556
Retail             450.236287
Manufacturing      421.818182
Sales              365.083333
Transportation     360.666667
Energy             352.857143
Construction       331.416667
Food               323.546584
Fitness            317.062500
Travel             316.533333
Aerospace          231.466667
Real Estate        212.845361
Name: total_laid_off, dtype: float64

# Key Insights

## 1. Finance recorded the highest number of layoff events.

The Finance industry reported 527 layoff events, the highest among all industries.
This indicates that workforce restructuring was widespread across financial organizations during the study period.

## 2. Retail and Hardware experienced the largest workforce reductions.

Retail companies laid off approximately 106,706 employees.
Hardware companies laid off approximately 106,157 employees.
These industries experienced the greatest overall workforce impact despite having fewer layoff events than Finance.

## 3. Hardware had the most severe layoffs.

Hardware recorded an average layoff size of approximately 1,930 employees per event, the highest among all industries.
This suggests that layoffs in the Hardware sector were less frequent but significantly larger when they occurred.

## 4. Consumer and Retail industries were consistently affected across multiple metrics.

Both industries ranked highly in:
Number of layoff events
Total employees laid off
Average layoff size
This indicates sustained workforce pressure across these sectors.

## 5. Industry impact differs depending on the metric used.

Finance led in layoff frequency.
Retail and Hardware led in total workforce reductions.
Hardware led in layoff severity (average employees affected per event).
Therefore, no single metric fully captures industry-level layoff risk.

## Question 3
Which countries were most affected by layoffs during the study period?

In [16]:
country_events = (
    df.groupby("country")
      .size()
      .sort_values(ascending=False)
)

country_events.head(15)

country
United States     2796
India              340
Canada             175
Israel             169
United Kingdom     152
Germany            134
Brazil              89
Australia           84
Singapore           55
Sweden              45
Indonesia           40
Nigeria             36
China               31
Netherlands         29
France              23
dtype: int64

In [17]:
country_layoffs = (
    df.groupby("country")["total_laid_off"]
      .sum()
      .sort_values(ascending=False)
)

country_layoffs.head(15)

country
United States     656271.0
India              66039.0
Germany            32055.0
United Kingdom     23354.0
Netherlands        21575.0
Sweden             20379.0
Canada             16123.0
Israel             14309.0
Brazil             11939.0
China               8190.0
Australia           7930.0
Singapore           7533.0
Indonesia           5711.0
Japan               4285.0
Kenya               3156.0
Name: total_laid_off, dtype: float64

In [18]:
country_summary = pd.DataFrame({
    "Layoff_Events": df.groupby("country").size(),
    "Employees_Laid_Off": df.groupby("country")["total_laid_off"].sum()
})

country_summary.sort_values(
    by="Employees_Laid_Off",
    ascending=False
).head(15)

,Layoff_Events,Employees_Laid_Off
country,,
United States,2796,656271.0
India,340,66039.0
Germany,134,32055.0
United Kingdom,152,23354.0
Netherlands,29,21575.0
Sweden,45,20379.0
Canada,175,16123.0
Israel,169,14309.0
Brazil,89,11939.0


In [19]:
country_avg = (
    df.groupby("country")["total_laid_off"]
      .mean()
      .sort_values(ascending=False)
)

country_avg.head(15)

country
Netherlands             1135.526316
Japan                   1071.250000
Sweden                   599.382353
China                    585.000000
Russia                   400.000000
Cayman Islands           373.333333
United States            369.938557
Kenya                    350.666667
Saudi Arabia             340.000000
Germany                  327.091837
Senegal                  300.000000
United Kingdom           274.752941
Indonesia                271.952381
Singapore                251.100000
United Arab Emirates     248.750000
Name: total_laid_off, dtype: float64

# Key Insights

## 1. The United States experienced the highest layoff activity by a significant margin.

The United States recorded 2,796 layoff events, the highest among all countries in the dataset.
It also accounted for approximately 656,271 employees laid off, making it the most affected country in terms of workforce reductions.

## 2. India ranked second in both layoff frequency and workforce impact.

India recorded 340 layoff events and approximately 66,039 employees laid off.
This indicates substantial workforce restructuring within India's startup and technology ecosystem during the study period.

## 3. Germany and the United Kingdom were the most affected European countries.

Germany recorded 134 layoff events and approximately 32,055 employees laid off.
The United Kingdom recorded 152 layoff events and approximately 23,354 employees laid off.

## 4. Workforce reductions were concentrated in a small number of countries.

The United States alone accounted for the majority of reported layoff events and employees affected.
A significant gap exists between the United States and all other countries, highlighting the concentration of layoffs in major startup and technology hubs.

## 5. Some countries experienced fewer layoff events but larger layoffs per event.

The Netherlands recorded an average of approximately 1,136 employees laid off per event.
Japan recorded an average of approximately 1,071 employees laid off per event.
This suggests that although layoffs were less frequent in these countries, workforce reductions were often larger when they occurred.

## Question 4
Which company stages are most vulnerable to layoffs?

In [20]:
stage_events = (
    df.groupby("stage")
      .size()
      .sort_values(ascending=False)
)

stage_events

stage
Post-IPO          1057
Unknown            749
Series B           484
Series C           445
Acquired           398
Series D           358
Series A           269
Series E           205
Seed               147
Series F           123
Private Equity      80
Series H            36
Subsidiary          29
Series G            28
Series J            11
Series I             9
dtype: int64

In [21]:
stage_layoffs = (
    df.groupby("stage")["total_laid_off"]
      .sum()
      .sort_values(ascending=False)
)

stage_layoffs

stage
Post-IPO          582668.0
Unknown            81183.0
Acquired           73178.0
Series B           32240.0
Series C           27530.0
Series D           27304.0
Series E           25314.0
Series F           16691.0
Private Equity     13080.0
Series A            9147.0
Series H            8497.0
Subsidiary          7538.0
Series G            5129.0
Series J            4700.0
Series I            3155.0
Seed                2751.0
Name: total_laid_off, dtype: float64

In [22]:
stage_summary = pd.DataFrame({
    "Layoff_Events": df.groupby("stage").size(),
    "Employees_Laid_Off": df.groupby("stage")["total_laid_off"].sum()
})

stage_summary.sort_values(
    by="Employees_Laid_Off",
    ascending=False
)

,Layoff_Events,Employees_Laid_Off
stage,,
Post-IPO,1057,582668.0
Unknown,749,81183.0
Acquired,398,73178.0
Series B,484,32240.0
Series C,445,27530.0
Series D,358,27304.0
Series E,205,25314.0
Series F,123,16691.0
Private Equity,80,13080.0


In [23]:
stage_avg = (
    df.groupby("stage")["total_laid_off"]
      .mean()
      .sort_values(ascending=False)
)

stage_avg

stage
Post-IPO          776.890667
Series J          470.000000
Series I          394.375000
Subsidiary        376.900000
Series H          283.233333
Acquired          279.305344
Private Equity    261.600000
Series G          244.238095
Series F          181.423913
Unknown           179.608407
Series E          163.316129
Series D          108.780876
Series B          104.336570
Series C           96.936620
Series A           57.892405
Seed               57.312500
Name: total_laid_off, dtype: float64

# Key Insights

## 1. Post-IPO companies experienced the highest layoff activity and workforce reductions.

Post-IPO firms recorded 1,057 layoff events, the highest among all company stages.
These companies accounted for approximately 582,668 employees laid off.
They also recorded the highest average layoff size (777 employees per event).
This indicates that publicly listed companies were the primary drivers of workforce reductions during the study period.

## 2. Growth-stage startups experienced frequent layoffs but at smaller scales.

Series B (484 events), Series C (445 events), and Series D (358 events) reported high layoff activity.
However, their average layoff sizes ranged between 97 and 108 employees per event, substantially lower than Post-IPO firms.
This suggests that growth-stage startups frequently adjusted workforce levels but typically conducted smaller layoffs.

## 3. Acquired companies underwent substantial restructuring.

Acquired firms recorded 398 layoff events and approximately 73,178 employees laid off.
Their average layoff size was approximately 279 employees per event, indicating that mergers and acquisitions were often accompanied by workforce consolidation.

## 4. Early-stage startups were the least affected.

Seed-stage companies recorded 147 layoff events and only 2,751 employees laid off.
Average layoff size was approximately 57 employees per event, the lowest among major stages.
This suggests that early-stage startups contributed relatively little to overall workforce reductions.

## 5. Layoff severity generally increased with company maturity.

Mature stages such as Post-IPO, Acquired, Private Equity, and Subsidiary companies reported significantly larger layoffs per event.
In contrast, Series A, Series B, Series C, and Seed companies reported much smaller workforce reductions on average.
This indicates that larger and more mature organizations tend to implement layoffs on a much greater scale when restructuring occurs.# 

## Question 5
Is there a relationship between funds raised and the magnitude of layoffs?

In [24]:
df["funds_raised"].describe()

count      3916.000000
mean        856.560494
std        4562.857311
min           0.700000
25%          55.000000
50%         175.500000
75%         483.250000
max      121900.000000
Name: funds_raised, dtype: float64

In [25]:
df[["funds_raised", "total_laid_off"]].corr()

,funds_raised,total_laid_off
funds_raised,1.00000,0.11917
total_laid_off,0.11917,1.00000


In [26]:
df[
    ["company", "funds_raised", "total_laid_off", "stage"]
].sort_values(
    by="funds_raised",
    ascending=False
).head(15)

,company,funds_raised,total_laid_off,stage
3041,Netflix,121900.0,30.0,Post-IPO
3462,Netflix,121900.0,300.0,Post-IPO
3687,Netflix,121900.0,25.0,Post-IPO
3651,Netflix,121900.0,150.0,Post-IPO
1127,Xerox,27200.0,3000.0,Post-IPO
10,Meta,26000.0,8000.0,Post-IPO
374,Meta,26000.0,100.0,Post-IPO
444,Meta,26000.0,3600.0,Post-IPO
588,Meta,26000.0,NaN,Post-IPO
76,Meta,26000.0,200.0,Post-IPO


In [27]:
df.groupby("stage")["funds_raised"] \
  .mean() \
  .sort_values(ascending=False)

stage
Series I          3715.111111
Series H          3091.972222
Subsidiary        3000.209091
Series J          2805.090909
Post-IPO          2206.823298
Series G           989.178571
Private Equity     842.181739
Series F           683.900000
Unknown            628.461639
Series E           622.687805
Acquired           499.267763
Series D           377.253143
Series C           204.858314
Series B           107.043235
Series A            44.110506
Seed                 7.013734
Name: funds_raised, dtype: float64

In [28]:
df["funding_group"] = pd.cut(
    df["funds_raised"],
    bins=[0, 50, 200, 500, 1000, 2000, 200000],
    labels=[
        "0-50M",
        "50-200M",
        "200-500M",
        "500M-1B",
        "1B-2B",
        "2B+"
    ]
)

In [29]:
funding_summary = df.groupby("funding_group").agg({
    "total_laid_off": "sum",
    "company": "count"
})

funding_summary

,total_laid_off,company
funding_group,,
0-50M,168731.0,934
50-200M,142238.0,1147
200-500M,97389.0,891
500M-1B,85149.0,469
1B-2B,97578.0,243
2B+,166199.0,232


# Key Insights

## 1. Funding raised showed only a weak relationship with layoffs.

The correlation between funds raised and employees laid off was approximately 0.12.
This indicates a very weak positive relationship.
Companies that raised more capital were not necessarily protected from workforce reductions.

## 2. Both highly funded and lightly funded companies conducted significant layoffs.

Companies that raised less than $50 million accounted for approximately 168,731 employees laid off.
Companies that raised more than $2 billion accounted for approximately 166,199 employees laid off.

This suggests that layoffs affected organizations across the entire funding spectrum.

## 3. The highest number of layoff events occurred among moderately funded companies.

Funding Group	Layoff Events
50–200M	           1,147
0–50M	            934
200–500M	        891
Mid-sized companies represented the largest share of layoff announcements.
These firms may have faced greater pressure to balance growth expectations with profitability goals.

## 4. Large funding rounds did not guarantee workforce stability.

Several highly funded companies, including major Post-IPO firms, still conducted large-scale layoffs.
This indicates that workforce reductions were often driven by strategic restructuring, market conditions, or profitability objectives rather than funding availability alone.

## 5. Layoffs were distributed across all funding categories.

Significant workforce reductions were observed in every funding bracket.
No funding category appeared immune to layoffs, suggesting that broader economic and business factors played a larger role than capital availability.

## Question 6
Which companies conducted the largest layoffs, and what characteristics do they share?

In [30]:
company_layoffs = (
    df.groupby("company")["total_laid_off"]
      .sum()
      .sort_values(ascending=False)
)

company_layoffs.head(20)

company
Amazon        58124.0
Intel         43115.0
Meta          35700.0
Oracle        31294.0
Microsoft     30055.0
Dell          23650.0
Cisco         18521.0
Salesforce    16525.0
Tesla         14500.0
Google        13697.0
SAP           11000.0
Ericsson      10100.0
Philips       10000.0
PayPal         9428.0
Flink          8100.0
HP             8100.0
Uber           7785.0
Micron         7500.0
Block          5931.0
Getir          5900.0
Name: total_laid_off, dtype: float64

In [31]:
company_summary = (
    df.groupby("company")
      .agg({
          "total_laid_off":"sum",
          "funds_raised":"first",
          "stage":"first",
          "industry":"first",
          "country":"first"
      })
      .sort_values(
          by="total_laid_off",
          ascending=False
      )
)

company_summary.head(20)

,total_laid_off,funds_raised,stage,industry,country
company,,,,,
Amazon,58124.0,8100.0,Post-IPO,Retail,United States
Intel,43115.0,12.0,Post-IPO,Hardware,United States
Meta,35700.0,26000.0,Post-IPO,Consumer,United States
Oracle,31294.0,NaN,Post-IPO,Other,United States
Microsoft,30055.0,1.0,Post-IPO,Other,United States
Dell,23650.0,NaN,Post-IPO,Hardware,United States
Cisco,18521.0,2.0,Post-IPO,Infrastructure,United States
Salesforce,16525.0,65.0,Post-IPO,Sales,United States
Tesla,14500.0,20200.0,Post-IPO,Transportation,United States


# Key Insights

## 1. Amazon recorded the highest number of employees laid off among all companies in the dataset, followed by Intel, Meta, Oracle, and Microsoft.

Amazon reported approximately 58,124 layoffs.
Intel reported approximately 43,115 layoffs.
Meta reported approximately 35,700 layoffs.
Oracle and Microsoft each reported more than 30,000 layoffs.

## 2. Most of the companies with the largest layoffs were Post-IPO organizations.

The majority of the top-ranked companies were publicly listed firms.
This indicates that mature organizations were more likely to implement large-scale workforce restructuring than early-stage startups.

## 3. The United States dominated the list of companies with the largest layoffs.

Most of the top companies, including Amazon, Meta, Microsoft, Oracle, Dell, Cisco, Google, Tesla, PayPal, Uber, and HP, were based in the United States.
This aligns with earlier findings that the United States experienced the highest overall layoff activity.

## 4. Technology-oriented industries were heavily represented among the largest layoffs.

Hardware companies such as Intel, Dell, HP, and Micron appeared frequently among the top layoff organizations.
Consumer and digital technology firms such as Meta and Google also reported substantial workforce reductions.
This suggests that the technology ecosystem experienced significant restructuring during the study period.

## 5. High funding levels did not prevent workforce reductions.

Several companies with substantial funding histories, including Meta, Tesla, Uber, and Amazon, still conducted major layoffs.
This reinforces the earlier finding that funding availability alone was not a strong predictor of workforce stability.

## 6. Large workforce reductions were concentrated among mature, global organizations.

Companies with the largest layoffs generally operated at a global scale, employed large workforces, and had established market positions.
Layoffs were therefore concentrated among mature enterprises rather than smaller startups.

In [32]:
df.shape

(4428, 16)

In [33]:
df.to_csv(
    "../outputs/cleaned_layoffs.csv",
    index=False
)